### PR TITLE
genpolicy: support darwin target 

### PR DESCRIPTION
**Pull request by burgerdev**
This PR allows building the genpolicy binary on darwin and adds a CI step to ensure it keeps building in the future.

Background: #10424. The individual commits have more detailed descriptions, too.

cc @Apokleos @Redent0r @danmihai1

### CODE REVIEW

File: src/libs/protocols/src/trans.rs [hunk 1]
```diff
@@ -41,8 +40,9 @@ fn cap_hashset2vec(hash_set: &Option<HashSet<oci::Capability>>) -> Vec<String> {
 fn cap_vec2hashset(caps: Vec<String>) -> HashSet<oci::Capability> {
     caps.iter()
         .map(|cap: &String| {
-            let cap_str = parse_json_string(cap);
-            cap_str
+            // cap might be JSON-encoded
+            let decoded: &str = serde_json::from_str(cap).unwrap_or(cap);
```
Comment (burgerdev) [inline src/libs/protocols/src/trans.rs:44]: `serde_json` needs to be declared.

**Review by Redent0r — APPROVED** [review]

**Comment by danmihai1** [comment]
@burgerdev , if I understand correctly, these would be the only CI tests for Darwin. And, I don't have a Darwin machine to debug in case something goes wrong. Therefore, I hope to get some guidance here from folks more experienced than me - e.g., @fidencio @wainersm @stevenhorsman.

**Comment by burgerdev** [comment]
Well, it's not the only CI test on Darwin - I'm adding a test case to the existing https://github.com/kata-containers/kata-containers/actions/workflows/darwin-tests.yaml. Fwiw, we do check compilation on other setups which are hard to come by, such as PowerPC or s390x. In those cases, we seem to rely on individual engineers with access to such setups (e.g. https://katacontainers.slack.com/archives/C879ACQ00/p1722528701289589). So, might we just rely on engineers with MacBooks for this case here?

**Comment by stevenhorsman** [comment]
Yeah from my perspective increasing the coverage of the darwin CI tests is okay, but I understand the point about Dan not being able to debug if there are arch specific issues, so I think it would fall back to the arch maintainers to have provide support. I'm not sure we have an Apple employees actively involved now (they introduced these ci tests IIRC), so we are probably working with an approach here that darwin only test failures are not required and if they become unstable for too long they are likely to get skipped.

**Comment by burgerdev** [comment]
cc @egernst (who added the Darwin workflow and seems to be active still)

@stevenhorsman: so, can we move forward here or do we need a maintainer to explicitly own this? I can also remove the CI integration from this PR, the remaining commits are still worth having imho.

**Comment by stevenhorsman** [comment]
> @stevenhorsman: so, can we move forward here or do we need a maintainer to explicitly own this? I can also remove the CI integration from this PR, the remaining commits are still worth having imho.

From my perspective, without a maintainer, darwin is an experimental/unstable feature. That doesn't mean that we can't still add enhancements/improvements to it, or extra CI tests, but means if/when it stops working it will be removed as a required test and the failing tests might be skipped if there is no-one available to fix them.

**Comment by danmihai1** [comment]
@burgerdev , @stevenhorsman . @Redent0r : my vote is to:
- Suggest to Darwin users to run genpolicy in a different environment - e.g., in a Linux container. BTW, this is currently our recommendation for Windows users too.
- Consider again adding Darwin support if enough users will request it.

**Comment by burgerdev** [comment]
Ok, point taken - I removed the CI changes from this PR. Can we still move forward with the other changes?

* [kata-sys-util: move json parsing to protocols crate](https://github.com/kata-containers/kata-containers/pull/10425/commits/dcc80085345f8d08f2a6324ac5670ed43ced0cb1)
* [protocols: only build RLimit impls on Linux/Solaris](https://github.com/kata-containers/kata-containers/pull/10425/commits/93ba894d68a332061345bf004d28afefcb8efc07)

**Comment by burgerdev** [comment]
Thanks, @fidencio, I'd be happy to maintain the Darwin build target for the developer tools. Until we get this sorted out, it would be nice if we could move forward with the current state of the PR, excluding CI. This should not present a maintenance burden yet and make some of the builds more lightweight (due to the lack of `kata-sys-utils` dependency).

**Comment by fidencio** [comment]
@burgerdev, please, just rebase your PR atop of `main` as I did a whole bunch of workflow changes that only will be taken into consideration after the rebase.

If the tests pass, I'll approve it, and then we discuss about the darwin client at some point in the future.

File: src/libs/protocols/src/trans.rs [hunk 2]
```diff
@@ -97,6 +97,7 @@ impl From<oci::LinuxCapabilities> for grpc::LinuxCapabilities {
     }
 }
 
+#[cfg(any(target_os = "linux", target_os = "solaris"))]
```
Comment (fidencio) [inline src/libs/protocols/src/trans.rs:100]: Do we have any link of the upstream issue for this?

We're not building for solaris, so I'd just mark "linux" as the target os.

File: src/libs/protocols/src/trans.rs [hunk 3]
```diff
@@ -118,6 +119,7 @@ impl From<oci::Process> for grpc::Process {
             Env: option_vec_to_vec(from.env()),
             Cwd: from.cwd().display().to_string(),
             Capabilities: from_option(from.capabilities().clone()),
+            #[cfg(any(target_os = "linux", target_os = "solaris"))]
```
Comment (fidencio) [inline src/libs/protocols/src/trans.rs:122]: ditto.

File: src/libs/protocols/src/trans.rs [hunk 4]
```diff
@@ -1078,6 +1081,8 @@ impl From<grpc::Process> for oci::Process {
         } else {
             process.set_capabilities(None);
         }
+
+        #[cfg(any(target_os = "linux", target_os = "solaris"))]
```
Comment (fidencio) [inline src/libs/protocols/src/trans.rs:1085]: ditto.

File: src/libs/protocols/src/trans.rs [hunk 5]
```diff
@@ -993,6 +995,7 @@ impl From<grpc::Linux> for oci::Linux {
     }
 }
 
+#[cfg(any(target_os = "linux", target_os = "solaris"))]
```
Comment (fidencio) [inline src/libs/protocols/src/trans.rs:998]: ditto.

**Comment by c3d** [comment]
> Yeah from my perspective increasing the coverage of the darwin CI tests is okay, but I understand the point about Dan not being able to debug if there are arch specific issues, so I think it would fall back to the arch maintainers to have provide support. I'm not sure we have an Apple employees actively involved now (they introduced these ci tests IIRC), so we are probably working with an approach here that darwin only test failures are not required and if they become unstable for too long they are likely to get skipped.

For what it's worth, @bpradipt and me (at least) have a Mac that we can use to test builds locally. So for minor breakage, that's relatively easy to deal with. And indeed, @egernst may have a vested interest in keeping this alive.

**Review by fidencio — APPROVED** [review]
lgtm, thanks @burgerdev!

This will get merged as soon as s390x CIs finish.

**Comment by burgerdev** [comment]
> @burgerdev The description states "_and adds a CI step to ensure it keeps building in the future_". I do not see the additional CI step in the changed files. I'm probably not familiar enough with how that part of the CI works.

@c3d:  Sorry, this part was removed after discussion (cf. https://github.com/kata-containers/kata-containers/pull/10425#issuecomment-2434509837). I updated the title but forgot to update the description.

**Pull request merged**